### PR TITLE
Document PGO in devguide 

### PR DESCRIPTION
**Pull request by KilJaeeun**
Hi! this is my first open source Pull Request in CPython Sprint at PyCon KR so it might be awkward. Added description of enable optimization option. Please take good care of me!
image

- Related issue
   - https://github.com/python/cpython/issues/88703

- I made this PullRequest with these references
   - https://docs.python.org/3/using/configure.html?highlight=lto#performance-options
   - https://github.com/python/cpython/blob/main/Doc/using/configure.rst#python-build-system
   - https://github.com/python/cpython/issues/88703

----
:books: Documentation preview :books:: https://cpython-devguide--1153.org.readthedocs.build/

### CODE REVIEW

**Comment by corona10** [comment]
@hugovk @CAM-Gerlach @AlexWaygood 

Dear reviewers, 
@KilJaeeun is a participant in the PyCon KR sprint. 
This is the first PR to the OSS world in her life ever!

**Review by AlexWaygood — COMMENTED** [review]
Thanks, it's really useful to document this! This looks pretty good.

Could you maybe start the section with a brief description of why you might or might not want to use this option? I worry that at the moment, people new to Python might think that this is the best option that they should use all the time (since "optimized" is a very positive adjective!)

Maybe the section could start with something like this?

> If you are trying to improve the performance of Python, you will probably want to use an optimized build of CPython. It can take a lot longer to build CPython with optimizations enabled, and it's usually not necessary to do so. However, it's essential if you want accurate benchmark results for a proposed performance optimization.

Could you also please wrap each line to 80 characters, like the other paragraphs in this document?

**Comment by AlexWaygood** [comment]
Closing and reopening to retrigger the CLA bot

File: getting-started/setup-building.rst [hunk 1]
```diff
@@ -237,6 +237,14 @@ build the :py:mod:`ctypes` module (without the flag the rest of CPython will
 still build properly).
 
 
+Optimization
+''''''''''''
+
+To get an optimized build of Python, ``configure --enable-optimizations --with-lto``. This sets the default make targets up to enable Profile Guided Optimization (PGO) and may be used to auto-enable Link Time Optimization (LTO) on some platforms. Note that configuration including this option takes time to configure. If you want to learn more about these options, please  see `Configure > Performance options <https://docs.python.org/3/using/configure.html?highlight=lto#performance-options>`_ ::
```
Comment (AA-Turner) [inline getting-started/setup-building.rst:243]: We can use intersphinx here, e.g. `` :ref:`performance options <python:performance options>` ``.

Also, we should use an explicit code block with the [``console`` highlighter](https://pygments.org/docs/lexers/#pygments.lexers.shell.BashSessionLexer), e.g.:

```restructuredtext
.. code:: console

   $ ./configure --enable-optimizations --with-lto
```

**Review by corona10 — DISMISSED** [review]
@KilJaeeun 

Would you like to update the PR with comments from other reviewers?

- https://github.com/python/devguide/pull/1153#pullrequestreview-1573385967
- https://github.com/python/devguide/pull/1153/files#r1291623444

**Comment by KilJaeeun** [comment]
@corona10 @AlexWaygood 
I reflected the code review Please check it out!

File: getting-started/setup-building.rst [hunk 2]
```diff
@@ -237,6 +237,25 @@ build the :py:mod:`ctypes` module (without the flag the rest of CPython will
...
+If you are trying to improve the performance of Python, you will probably want
+to use an optimized build of CPython. It can take a lot longer to build CPython
+with optimizations enabled, and it's usually not necessary to do so. However,
+it's essential if you want accurate benchmark results for a proposed performance
+optimization. To get an optimized build of Python,
+``configure --enable-optimizations --with-lto``.
+This sets the default make targets up to enable Profile Guided Optimization(PGO)
+and may be used to auto-enable Link Time Optimization (LTO) on some platforms.
+If you want to learn more about these options,
+please  see :ref:`Configure options <python:configure-options>`.
```
Comment (AlexWaygood) [inline getting-started/setup-building.rst:252]: Slightly more concise, and just as polite:

```suggestion
See :ref:`Configure options <python:configure-options>` to learn more about
these options.
```

**Review by AlexWaygood — COMMENTED** [review]
Thanks, nearly there! A few more small suggestions below. If you like my suggestions, you can apply them by clicking the "commit suggestion" button below each suggestion :)

File: getting-started/setup-building.rst [hunk 3]
```diff
@@ -237,6 +237,25 @@ build the :py:mod:`ctypes` module (without the flag the rest of CPython will
 still build properly).
 
 
+Optimization
+''''''''''''
+
+If you are trying to improve the performance of Python, you will probably want
```
Comment (AA-Turner) [inline getting-started/setup-building.rst:243]:
```suggestion
If you are trying to improve CPython's performance, you will probably want
```

File: getting-started/setup-building.rst (hunk 2 above)
Comment (AA-Turner) [reply getting-started/setup-building.rst:252]: Untested, but you may be able to directly link to the performance options section:

```suggestion
See :ref:`Configure options <python:performance options>` to learn more about
these options.
```

**Review by AA-Turner — COMMENTED** [review]
Depending on how much detail you want to go into, you could link the specific options themselves with `` :option`python:--enable-optimizations` `` or `` :option`python:--with-lto` ``.


--- 

Related, I was a little surprised to see [*configure options*](https://docs.python.org/3/using/configure.html) in d.p.o -- thinking about it, should we consider moving it to the devguide? I'm not sure many end-users build their own Python from sources... 

A

File: getting-started/setup-building.rst (hunk 2 above)
Comment (AlexWaygood) [reply getting-started/setup-building.rst]: Looks like Adam's suggestion here is causing the build to fail, so I'm changing it back to

```
:ref:`Configure options <python:configure-options>`
```

**Review by AlexWaygood — APPROVED** [review]
Thanks, looks great!

File: getting-started/setup-building.rst (hunk 2 above)
Comment (AA-Turner) [reply getting-started/setup-building.rst]: Thanks Alex!

File: .idea/workspace.xml
Comment (AA-Turner) [inline .idea/workspace.xml:1]: I think this file was accidentally committed @KilJaeeun

Comment (AlexWaygood) [reply .idea/workspace.xml:1]: I already removed it :)

File: getting-started/setup-building.rst [hunk 4]
```diff
@@ -240,18 +240,18 @@ still build properly).
...
 
 For an optimized build of Python, use
 ``configure --enable-optimizations --with-lto``.
-This sets the default make targets up to enable Profile Guided Optimization(PGO)
+This sets the default make targets up to enable Profile Guided Optimization (PGO)
 and may be used to auto-enable Link Time Optimization (LTO) on some platforms.
-If you want to learn more about these options,
-please  see :ref:`Configure options <python:configure-options>`.
+See :option:`python:--enable-optimizations`, :option:`python:--with-lto` to learn more about
+these options.
```
Comment (AA-Turner) [inline getting-started/setup-building.rst:254]: To keep the line length shorter and a small grammar suggestion:
```suggestion
See :option:`python:--enable-optimizations` and :option:`python:--with-lto`
to learn more about these options.
```

**Review by AA-Turner — APPROVED** [review]
Looks good overall! Thanks @KilJaeeun and again congratulations on your first PR!

A

**Comment by AlexWaygood** [comment]
Thanks again @KilJaeeun, this was a great first PR!

**Pull request merged**